finality-aleph/src/network/manager/discovery.rs

```diff
@@ -31,38 +32,51 @@ impl DiscoveryMessage {
 }
 
 /// Handles creating and responding to discovery messages.
-pub struct Discovery {
+pub struct Discovery<M: Multiaddress> {
     cooldown: Duration,
     last_broadcast: HashMap<NodeIndex, Instant>,
+    _phantom: PhantomData<M>,
 }
 
-type DiscoveryCommand = (DiscoveryMessage, DataCommand);
+type DiscoveryCommand<M> = (
+    DiscoveryMessage<M>,
+    DataCommand<<M as Multiaddress>::PeerId>,
+);
 
-fn authentication_broadcast(authentication: Authentication) -> DiscoveryCommand {
+fn authentication_broadcast<M: Multiaddress>(
+    authentication: Authentication<M>,
+) -> DiscoveryCommand<M> {
     (
         DiscoveryMessage::AuthenticationBroadcast(authentication),
         DataCommand::Broadcast,
     )
 }
 
-fn response(authentication: Authentication, peer_id: PeerId) -> DiscoveryCommand {
+fn response<M: Multiaddress>(
+    authentication: Authentication<M>,
+    peer_id: M::PeerId,
+) -> DiscoveryCommand<M> {
     (
         DiscoveryMessage::Authentication(authentication),
         DataCommand::SendTo(peer_id, Protocol::Generic),
     )
 }
 
-impl Discovery {
+impl<M: Multiaddress> Discovery<M> {
     /// Create a new discovery handler with the given response/broadcast cooldown.
     pub fn new(cooldown: Duration) -> Self {
         Discovery {
             cooldown,
             last_broadcast: HashMap::new(),
+            _phantom: PhantomData,
```

Comment:
why do you need PhantomData here if it doesn't hold a generic?

Reply:
It holds `M: Multiaddress`, as written in the struct definition. I think I could rewrite this whole thing to only have the methods parametrized by `M`, but this would be less clear imo.